File: frb-compare.cpp
#include <mpi.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <fstream>
#include "frb_olympics.hpp"

using namespace std;
using namespace frb_olympics;


// -------------------------------------------------------------------------------------------------
//
// Node info (used in memhack)


static int mpi_tasks_per_node = 0;
static int mpi_rank_within_node = 0;


static void init_node_info()
{
    if (mpi_tasks_per_node > 0)
	return;   // already initialized

    int mpi_rank = MPI::COMM_WORLD.Get_rank();
    int mpi_size = MPI::COMM_WORLD.Get_size();

    vector<char> hname_buf1(HOST_NAME_MAX+2, 0);
    vector<char> hname_buf2(mpi_size * (HOST_NAME_MAX+2), 0);

    if (gethostname(&hname_buf1[0], HOST_NAME_MAX+1) < 0)
	throw runtime_error("rng_initializer: gethostname() failed");

    MPI::COMM_WORLD.Allgather(&hname_buf1[0], HOST_NAME_MAX+2, MPI::CHAR,
			      &hname_buf2[0], HOST_NAME_MAX+2, MPI::CHAR);

    string my_hname = string(&hname_buf1[0]);
    boost::unordered_map<string,int> hname_counts;

    for (int i = 0; i < mpi_size; i++) {
	string s = string(&hname_buf2[i * (HOST_NAME_MAX+2)]);

	if (s == my_hname) {
	    mpi_tasks_per_node++;
	    if (i < mpi_rank)
		mpi_rank_within_node++;
	}

	boost::unordered_map<string,int>::iterator p = hname_counts.find(s);
	int c = (p == hname_counts.end()) ? 0 : p->second;
	hname_counts[s] = c+1;
    }

    // check
    for (boost::unordered_map<string,int>::iterator p = hname_counts.begin(); p != hname_counts.end(); p++) {
	if (p->second != mpi_tasks_per_node)
	    throw runtime_error("init_node_info: expected all hostnames to have equal numbers of MPI tasks");
    }

    cout << "rank_within_node=" << mpi_rank_within_node << ", tasks_per_node=" << mpi_tasks_per_node << endl;
}


static void memhack_start(int memhack)
{
    if (memhack == 1)
	return;

    init_node_info();   // no-ops if already initialized
    xassert(memhack >= 1);
    xassert(mpi_tasks_per_node % memhack == 0);

    int nbarriers = (mpi_rank_within_node % memhack) + 1;
    for (int i = 0; i < nbarriers; i++)
	MPI::COMM_WORLD.Barrier();
}


static void memhack_end(int memhack)
{
    if (memhack == 1)
	return;

    init_node_info();   // no-ops if already initialized
    xassert(memhack >= 1);
    xassert(mpi_tasks_per_node % memhack == 0);

    int nbarriers = memhack - (mpi_rank_within_node % memhack);
    for (int i = 0; i < nbarriers; i++)
	MPI::COMM_WORLD.Barrier();
}


// -------------------------------------------------------------------------------------------------


static void usage()
{
    cerr << "usage: frb-compare <search_params.txt> <algo_list.txt> <nmc_noise> <nmc_pulse> <output_stem> [target_sn]\n"
	 << "       target_sn defaults to 30.0\n"
	 << "\n";

    frb_search_algorithm_base::show_algorithm_usage(cerr);
    exit(2);
}


int main(int argc, char **argv)
{
    MPI::Init(argc, argv);

    if ((argc < 6) || (argc > 7))
	usage();

    int mpi_rank = MPI::COMM_WORLD.Get_rank();
    int mpi_size = MPI::COMM_WORLD.Get_size();
    
    const char *search_params_filename = argv[1];
    const char *algo_list_filename = argv[2];
    int nmc_noise_tot = xlexical_cast<int> (argv[3], "frb-compare command line");
    int nmc_pulse_tot = xlexical_cast<int> (argv[4], "frb-compare command line");
    const char *output_stem = argv[5];
    const double target_sn = (argc == 7) ? xlexical_cast<double>(argv[6],"target_sn") : 30.0;

    // Redirect stdout to log file which is unique to this MPI rank
    if (mpi_size > 1) {
	stringstream slog;
	slog << output_stem << "_log" << mpi_rank;

	string log_filename = slog.str();
	cout << "redirecting stdout -> " << log_filename << endl;

	int log_fd = open(log_filename.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0644);
	
	if (log_fd < 0) {
	    cerr << "couldn't create log file " << log_filename << ": " << strerror(errno) << endl;
	    exit(1);
	}

	// stdout is file descriptor 1
	if (dup2(log_fd, 1) < 0)
	    cerr << "dup2() failed: " << strerror(errno) << endl;
    }

    cout << "command line:";
    for (int i = 0; i < argc; i++) 
	cout << " " << argv[i];
    cout << endl;

    xassert(nmc_noise_tot >= 0);
    xassert(nmc_pulse_tot >= 0);
    xassert(nmc_noise_tot % mpi_size == 0);
    xassert(nmc_pulse_tot % mpi_size == 0);

    int nmc_noise_loc = nmc_noise_tot / mpi_size;
    int nmc_pulse_loc = nmc_pulse_tot / mpi_size;

    frb_rng rng;
    frb_search_params search_params(search_params_filename);

    vector<frb_search_algorithm_base::ptr_t> algo_list;
    frb_search_algorithm_base::read_file(algo_list_filename, search_params, algo_list);
    int nalgo = algo_list.size();

    ofstream f_noise;
    ofstream f_pulse;

    if ((mpi_rank == 0) && (nmc_noise_tot + nmc_pulse_tot > 0)) {
	//
	// Open output files
	// Note: filename convention must be kept in sync with frb-compare-postprocess.py
	//
	// FIXME should move files out of the way if they already exist (currently silently clobbers)
	//
	stringstream s1;
	s1 << output_stem << "_noise.txt";

	string filename1 = s1.str();
	f_noise.open(filename1.c_str());

	if (f_noise.fail()) {
	    cerr << "Fatal: couldn't open file " << filename1 << endl;
	    exit(1);
	}

	stringstream s2;
	s2 << output_stem << "_pulse.txt";

	string filename2 = s2.str();
	f_pulse.open(filename2.c_str());

	if (f_pulse.fail()) {
	    cerr << "Fatal: couldn't open file " << filename2 << endl;
	    exit(1);
	}
    }

    cout << "--------------------  Search params  --------------------\n";
    search_params.write(cout);
    cout << endl;

    cout << "--------------------  Memory analysis  --------------------\n";

    double timestream_gb = 1.0e-9 * search_params.nchan * search_params.nsamples_per_chunk * sizeof(float);
    cout << "timestream chunk size = " << timestream_gb << " GB\n";

    double search_gb = 0.0;
    for (int ialgo = 0; ialgo < nalgo; ialgo++) {
	frb_search_algorithm_base::ptr_t algo = algo_list[ialgo];
	cout << "    " << algo->name << " " << algo->search_gb << " GB\n";
	search_gb = max(search_gb, algo->search_gb);
    }

    cout << "total size = " << (timestream_gb + search_gb) << " GB\n";

    vector<double> score_noise(nmc_noise_loc * nalgo);
    vector<double> score_pulse(nmc_pulse_loc * nalgo);
    vector<double> pulse_tab(nmc_pulse_loc * 5);

    // --------------------  noise sims  --------------------

    for (int inoise = 0; inoise < nmc_noise_loc; inoise++) {
	cout << "starting noise sim " << inoise << "/" << nmc_noise_loc << endl;

	vector<float> chunk(search_params.nchan * search_params.nsamples_per_chunk);
	frb_rng rsave(rng);   // save

	for (int ialgo = 0; ialgo < nalgo; ialgo++) {
	    frb_search_algorithm_base::ptr_t algo = algo_list[ialgo];
	    struct timeval tv0 = get_time();

	    memhack_start(algo->memhack);
	    algo->search_start();

	    rng = rsave;     // restore
	    for (int ichunk = 0; ichunk < search_params.nchunks; ichunk++) {
		cout << "     " << algo->name 
		     << ": starting chunk " << ichunk << "/" << search_params.nchunks 
		     << "  [time=" << time_diff(tv0,get_time()) << " secs]" << endl;

		search_params.simulate_noise(rng, &chunk[0]);
		algo->search_chunk(&chunk[0], ichunk, NULL);
	    }

	    cout << "     " << algo->name 
		 << ": search_result = " << algo->search_result 
		 << "  [time=" << time_diff(tv0,get_time()) << " secs]" << endl;

	    score_noise[inoise*nalgo + ialgo] = algo->search_result;

	    algo->search_end();
	    memhack_end(algo->memhack);
	}
    }

    // --------------------  pulse sims  --------------------

    for (int ipulse = 0; ipulse < nmc_pulse_loc; ipulse++) {
	cout << "starting pulse sim " << ipulse << "/" << nmc_pulse_loc << endl;

	frb_pulse p = search_params.make_random_pulse(rng, 1.0);
	pulse_tab[5*ipulse] = p.arrival_time;
	pulse_tab[5*ipulse+1] = p.intrinsic_width;
	pulse_tab[5*ipulse+2] = p.dispersion_measure;
	pulse_tab[5*ipulse+3] = p.scattering_measure;
	pulse_tab[5*ipulse+4] = p.spectral_index;

	double sn = search_params.get_signal_to_noise_of_pulse(p);
	p.fluence *= (target_sn / sn);
	
	vector<float> chunk(search_params.nchan * search_params.nsamples_per_chunk);	

	for (int ialgo = 0; ialgo < nalgo; ialgo++) {
	    frb_search_algorithm_base::ptr_t algo = algo_list[ialgo];	    
	    struct timeval tv0 = get_time();

	    memhack_start(algo->memhack);
	    algo->search_start();

	    for (int ichunk = 0; ichunk < search_params.nchunks; ichunk++) {
		cout << "     " << algo->name 
		     << ": starting chunk " << ichunk << "/" << search_params.nchunks 
		     << "  [time=" << time_diff(tv0,get_time()) << " secs]" << endl;

		memset(&chunk[0], 0, chunk.size() * sizeof(chunk[0]));
		search_params.add_pulse(p, &chunk[0], ichunk);
		algo->search_chunk(&chunk[0], ichunk, NULL);
	    }

	    cout << "     " << algo->name 
		 << ": search_result = " << algo->search_result 
		 << "  [time=" << time_diff(tv0,get_time()) << " secs]" << endl;
	    
	    score_pulse[ipulse*nalgo + ialgo] = algo->search_result;

	    algo->search_end();
	    memhack_end(algo->memhack);
	}
    }

    vector<double> score_noise_all(nmc_noise_tot * nalgo, 0.0);
    vector<double> score_pulse_all(nmc_pulse_tot * nalgo);
    vector<double> pulse_tab_all(nmc_pulse_tot * 5, 0.0);

    MPI::COMM_WORLD.Gather(&score_noise[0], nmc_noise_loc * nalgo, MPI::DOUBLE,
			   &score_noise_all[0], nmc_noise_loc * nalgo, MPI::DOUBLE, 0);

    MPI::COMM_WORLD.Gather(&score_pulse[0], nmc_pulse_loc * nalgo, MPI::DOUBLE,
			   &score_pulse_all[0], nmc_pulse_loc * nalgo, MPI::DOUBLE, 0);

    MPI::COMM_WORLD.Gather(&pulse_tab[0], 5 * nmc_pulse_loc, MPI::DOUBLE,
			   &pulse_tab_all[0], 5 * nmc_pulse_loc, MPI::DOUBLE, 0);

    if (mpi_rank == 0) {
	for (int i = 0; i < nmc_noise_tot; i++) {
	    for (int j = 0; j < nalgo; j++)
		f_noise << " " << score_noise_all[i*nalgo+j];
	    f_noise << "\n";
	}

	for (int i = 0; i < nmc_pulse_tot; i++) {
	    for (int j = 0; j < 5; j++)
		f_pulse << " " << pulse_tab_all[5*i+j];
	    for (int j = 0; j < nalgo; j++)
		f_pulse << " " << score_pulse_all[i*nalgo+j];
	    f_pulse << "\n";
	}
    }

    MPI::Finalize();
    return 0;
}

File: frb-compare.py
#!/usr/bin/env python

import os
import sys
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--sn', type=float, default=30.0)
parser.add_argument('-o', '--outstem')
parser.add_argument('search_params_txtfile')
parser.add_argument('algo_pyfile')
parser.add_argument('nmc_noise', type=int)
parser.add_argument('nmc_pulse', type=int)

args = parser.parse_args()

if args.outstem is not None:
    output_stem = args.outstem
else:
    assert args.algo_pyfile[-3:] == '.py'
    assert len(args.algo_pyfile) >= 4
    output_stem = args.algo_pyfile[:-3]

assert args.sn > 0.0
assert args.nmc_noise >= 0
assert args.nmc_pulse >= 0


####################################################################################################


import time
import numpy as np
import frb_olympics

frb_olympics.init_mpi_log_files(output_stem)

print 'command line: %s' % (' '.join(sys.argv))
print 'mpi_rank=%d, mpi_size=%d' % (frb_olympics.mpi_rank, frb_olympics.mpi_size)
print 'mpi_rank_within_node=%d, mpi_tasks_per_node=%d' % (frb_olympics.mpi_rank_within_node, frb_olympics.mpi_tasks_per_node)

assert args.nmc_noise % frb_olympics.mpi_size == 0
assert args.nmc_pulse % frb_olympics.mpi_size == 0
(nmc_noise_tot, nmc_pulse_tot) = (args.nmc_noise, args.nmc_pulse)

nmc_noise_loc = nmc_noise_tot // frb_olympics.mpi_size
nmc_pulse_loc = nmc_pulse_tot // frb_olympics.mpi_size


search_params = frb_olympics.frb_search_params(args.search_params_txtfile)
frb_olympics.imp(args.algo_pyfile)
frb_olympics.init_algorithms(search_params)
nalgo = len(frb_olympics.algo_list)

print '--------------------  Search params  --------------------'
search_params.write()

print '--------------------  Memory analysis  --------------------'

timestream_gb = 1.0e-9 * search_params.nchan * search_params.nsamples_per_chunk * 4.0
print 'timestream chunk size = %s GB' % timestream_gb

search_gb = 0.0
for algo in frb_olympics.algo_list:
    print '    %s %s GB' % (algo.name, algo.search_gb)
    search_gb = max(search_gb, algo.search_gb)

print 'total size = %s GB' % (timestream_gb + search_gb)

chunk = np.zeros((search_params.nchan,search_params.nsamples_per_chunk), dtype=np.float32)
score_noise = np.zeros((nmc_noise_loc,nalgo), dtype=np.float)
score_pulse = np.zeros((nmc_pulse_loc,nalgo), dtype=np.float)
pulse_tab = np.zeros((nmc_pulse_loc,5), dtype=np.float)
rng = frb_olympics.frb_rng()


for inoise in xrange(nmc_noise_loc):
    print 'starting noise sim %d/%d' % (inoise, nmc_noise_loc)
    rsave = frb_olympics.frb_rng(rng)

    for (ialgo, algo) in frb_olympics.enumerate_algorithms_with_memhack():
        rng = frb_olympics.frb_rng(rsave)
        t0 = time.time()

        for ichunk in xrange(search_params.nchunks):
            print '    %s: starting chunk %s/%s' % (algo.name, ichunk, search_params.nchunks)
            search_params.simulate_noise(rng, chunk)
            algo.search_chunk(chunk, ichunk)

        print '    %s: search_result=%s [time=%s]' % (algo.name, algo.search_result, time.time()-t0)
        score_noise[inoise,ialgo] = algo.search_result


for ipulse in xrange(nmc_pulse_loc):
    print 'starting pulse sim %d/%d' % (ipulse, nmc_pulse_loc)

    p = search_params.make_random_pulse(rng, 1.0)
    p.fluence = args.sn / search_params.get_signal_to_noise_of_pulse(p)

    pulse_tab[ipulse,0] = p.arrival_time
    pulse_tab[ipulse,1] = p.intrinsic_width
    pulse_tab[ipulse,2] = p.dispersion_measure
    pulse_tab[ipulse,3] = p.scattering_measure
    pulse_tab[ipulse,4] = p.spectral_index

    for (ialgo, algo) in frb_olympics.enumerate_algorithms_with_memhack():
        t0 = time.time()

        for ichunk in xrange(search_params.nchunks):
            print '    %s: starting chunk %s/%s' % (algo.name, ichunk, search_params.nchunks)
            chunk[:,:] = 0.0
            search_params.add_pulse(p, chunk, ichunk)
            algo.search_chunk(chunk, ichunk)

        print '    %s: search_result=%s [time=%s]' % (algo.name, algo.search_result, time.time()-t0)
        score_pulse[ipulse,ialgo] = algo.search_result


# lists of numpy arrays
score_noise = frb_olympics.mpi_gather(score_noise)
score_pulse = frb_olympics.mpi_gather(score_pulse)
pulse_tab = frb_olympics.mpi_gather(pulse_tab)


if frb_olympics.mpi_rank == 0:
    noise_data = np.concatenate(score_noise)
    noise_filename = args.outstem + '_noise.txt'
    np.savetxt(noise_filename, noise_data)
    print 'wrote %s' % noise_filename

    pulse_data = np.concatenate((np.concatenate(pulse_tab), np.concatenate(score_pulse)), axis=1)
    pulse_filename = args.outstem + '_pulse.txt'

    f = open(pulse_filename, 'w')
    print >>f, '# col 0: arrival time'
    print >>f, '# col 1: intrinsic width'
    print >>f, '# col 2: dispersion measure'
    print >>f, '# col 3: scattering measure'
    print >>f, '# col 4: spectral index'

    np.savetxt(f, pulse_data)
    print 'wrote %s' % pulse_filename

    import matplotlib
    matplotlib.use('Agg')

    c = frb_olympics.comparison_outputs(noise_data, pulse_data)
    c.plot_sigma(args.outstem)
